### ISSUE TITLE
SIMD unit test failures on BDW arch (AVX2) with gcc/9.3.0, 10.2.0

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#6277]
**Describe the bug**
The `Kokkos_UnitTest_SIMD` unit tests fail on BDW arch with gcc/9.3.0 and gcc/10.2.0 (other versions of gcc were not available on the system for testing):

**gcc/9.3.0 failure snip**:
```
33: [ RUN      ] simd.host
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 13
33:   b
33:     Which is: 1
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 3
33:   b
33:     Which is: 2
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 2
33:   b
33:     Which is: 1
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 0
33:   b
33:     Which is: 10
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 32-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00>
33:   b
33:     Which is: 32-byte object <FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF>
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 13
33:   b
33:     Which is: 0
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 3
33:   b
33:     Which is: 1
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 0
33:   b
33:     Which is: 10
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 32-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 FF-FF FF-FF FF-FF FF-FF 00-00 00-00 00-00 00-00>
33:   b
33:     Which is: 32-byte object <FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF>
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 13
33:   b
33:     Which is: 0
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 3
33:   b
33:     Which is: 1
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 32-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 FF-FF FF-FF FF-FF FF-FF 00-00 00-00 00-00 00-00>
33:   b
33:     Which is: 32-byte object <FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF 00-00 00-00 00-00 00-00>
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 13
33:   b
33:     Which is: 1
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 3
33:   b
33:     Which is: 2
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 2
33:   b
33:     Which is: 1
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 0
33:   b
33:     Which is: 10
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 32-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00>
33:   b
33:     Which is: 32-byte object <FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF>
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 13
33:   b
33:     Which is: 0
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 3
33:   b
33:     Which is: 1
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 0
33:   b
33:     Which is: 10
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 32-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 FF-FF FF-FF FF-FF FF-FF 00-00 00-00 00-00 00-00>
33:   b
33:     Which is: 32-byte object <FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF>
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 13
33:   b
33:     Which is: 0
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 3
33:   b
33:     Which is: 1
33: /home/ndellin/kokkos/simd/unit_tests/TestSIMD.cpp:26: Failure
33: Expected equality of these values:
33:   a
33:     Which is: 32-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 FF-FF FF-FF FF-FF FF-FF 00-00 00-00 00-00 00-00>
33:   b
33:     Which is: 32-byte object <FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF FF-FF 00-00 00-00 00-00 00-00>
33: [  FAILED  ] simd.host (1 ms)
33: [ RUN      ] simd.device
33: [       OK ] simd.device (1 ms)
33: [----------] 2 tests from simd (2 ms total)
33: 
33: [----------] Global test environment tear-down
33: [==========] 2 tests from 1 test suite ran. (2 ms total)
33: [  PASSED  ] 1 test.
33: [  FAILED  ] 1 test, listed below:
33: [  FAILED  ] simd.host
33: 
33:  1 FAILED TEST
1/1 Test #33: Kokkos_UnitTest_SIMD .............***Failed  Error regular expression found in output. Regex=[  FAILED  ]  0.01 sec
```

This occurred on a system with Broadwell arch (Intel® Xeon® Processor E5-2698 v4)

I also tested on a system with Skylake arch but setting `Kokkos_ARCH_BDW=ON` (SKX supports avx2) and encountered similar failures

**Please also include the following items to support reproducing the bug**
**Reproducer** (Serial build):
```
module load cmake/3.19 gcc/9.3.0

cmake -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_SERIAL=ON -DKokkos_ARCH_BDW=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_BUILD_TYPE=RELEASE -DKokkos_ENABLE_DEPRECATION_WARNINGS=OFF -DKokkos_ENABLE_DEPRECATED_CODE_4=OFF $KOKKOS_PATH
```

**Comment by ndellingwood** [comment]
@ldh4 I ran the tests on a machine with BDW where I use kokkos for another project
I also tested on Blake with gcc/8.3.0 and using `Kokkos_ARCH_BDW=ON` and saw similar errors; Blake has SKX arch which supports avx2, but I don't know that testing in this way (using `Kokkos_ARCH_BDW=ON` on SKX arch) is reliable for the SIMD tests?
I am looking into an account for testing on Solo, which is on the OHPC and has BDW arch. If you are unable to reproduce the failures maybe it is best to revisit this after I have testing in place on Solo and verify it reproduces on that machine

**Comment by masterleinad** [comment]
Looks like https://github.com/kokkos/kokkos/pull/6188 isn't detecting anything either.

**Comment by ndellingwood** [comment]
@ldh4 I reproduced the test failures on Solo with gcc/9.3.1, are you able to access that machine?

**Comment by ldh4** [comment]
@ndellingwood It doesn't look like I can access Solo. I have access to other Broadwell machines, but I still couldn't reproduce this issue there either. Do you know where I can request access to Solo?

**Comment by ndellingwood** [comment]
@ldh4 you can request "ECN Capacity Clusters" on WebCARS to gain access to solo. Login requires use of ECN password. Use of compute nodes requires being a user on a wcid which I don't yet have the means to provide, but if you build with the Serial backend with minimal or no parallelism on the login node you would be able to run an initial test there for the time being

**Comment by ldh4** [comment]
I got to reproduce this on Solo. Despite the test being serial, the failure seems indeterministic. The problem shows up for unary op test using `load_masked` as host loader. But if provided sufficient delay after the load (using `printf` for instance), then all tests pass as expected.
It's unclear to me why this is, but I'll look into it.

**Comment by masterleinad** [comment]
`gcc-8.4.0` is complaining about potentially uninitialized values.

**Comment by masterleinad** [comment]
Fixed in #6294.
